File: src/preflight.js
// Preflight — the pre-demo ritual as a button rather than a printed checklist.
//
// docs/14-TEST-PLAN.md §6 lists ten things to verify before every judging round. Under sleep
// deprivation at hour 26, a checklist you have to remember to run is a checklist you skip. This
// runs it in about two seconds and tells you exactly what is wrong.
//
// The debug-overlay check is here because it is the single easiest thing to forget, and a debug
// overlay on screen in front of a jury reads as unfinished.

export const Status = { PASS: 'PASS', WARN: 'WARN', FAIL: 'FAIL', SKIP: 'SKIP' };

/**
 * @param ctx { store, engine, audio, haptics, voice, speech, running, arenaMode, duel }
 * @returns {Promise<Array<{id,label,status,detail}>>}
 */
export async function preflight(ctx = {}) {
  const out = [];
  const add = (id, label, status, detail = '') => out.push({ id, label, status, detail });

  // 1 · camera actually delivering frames
  if (!ctx.running) add('camera', 'Camera running', Status.FAIL, 'Start the camera before demoing.');
  else if ((ctx.fps ?? 0) < 12) add('camera', 'Camera running', Status.WARN, `only ${Math.round(ctx.fps ?? 0)} fps`);
  else add('camera', 'Camera running', Status.PASS, `${Math.round(ctx.fps)} fps`);

  // 2 · inference inside the frame budget
  const infer = ctx.inferMs ?? null;
  if (infer === null) add('infer', 'Inference latency', Status.SKIP, 'no frames yet');
  else if (infer > 40) add('infer', 'Inference latency', Status.FAIL, `${infer.toFixed(0)}ms — likely on CPU, not GPU`);
  else if (infer > 22) add('infer', 'Inference latency', Status.WARN, `${infer.toFixed(0)}ms, budget is 22`);
  else add('infer', 'Inference latency', Status.PASS, `${infer.toFixed(0)}ms`);

  // 3 · config is the tuned version, not a debug variant
  const pose = ctx.store?.pose;
  if (!pose) add('config', 'Config loaded', Status.FAIL, 'no pose config');
  else if (pose.debugOverlay) add('config', 'Debug overlay OFF', Status.FAIL, 'debugOverlay is true in pose.json — a jury will see it');
  else add('config', 'Config loaded, debug off', Status.PASS, `v${pose.version ?? 1}`);

  // 4 · the exercise library resolved
  const n = Object.keys(ctx.store?.exercises ?? {}).length;
  add('library', 'Exercise library', n > 0 ? Status.PASS : Status.FAIL, `${n} exercises`);

  // 5 · offline. The claim is checkable, so check it.
  const online = typeof navigator !== 'undefined' ? navigator.onLine !== false : null;
  if (online === null) add('offline', 'Works offline', Status.SKIP, '');
  else add('offline', 'Works offline', online ? Status.WARN : Status.PASS,
           online ? 'currently online — put it in airplane mode and confirm the fight still runs' : 'offline and running');

  // 6 · audio unlocked (browsers need a gesture first)
  const actx = ctx.audio?.ctx;
  if (!actx) add('audio', 'Audio ready', Status.WARN, 'not unlocked yet — click anything once');
  else add('audio', 'Audio ready', actx.state === 'running' ? Status.PASS : Status.WARN, actx.state);

  // 7 · speech
  add('speech', 'Speech available', ctx.speech?.enabled ? Status.PASS : Status.WARN,
      ctx.speech?.enabled ? (ctx.speech.voice?.lang ?? 'default voice') : 'no speech synthesis');

  // 8 · haptics
  add('haptics', 'Haptics', ctx.haptics?.available ? Status.PASS : Status.SKIP,
      ctx.haptics?.available ? (ctx.haptics.enabled ? 'on' : 'off') : 'no vibration motor');

  // 9 · storage survives a write
  const store = ctx.store2;
  if (!store) add('storage', 'Local storage', Status.SKIP, '');
  else {
    const before = store.sessions.length;
    add('storage', 'Local storage', Status.PASS, `${before} sessions, streak ${store.streak.current}`);
  }

  // 10 · calibration for the selected exercise
  const cal = ctx.store2?.calibrationFor?.(ctx.exerciseId);
  add('calibration', 'Calibration for this exercise', cal ? Status.PASS : Status.WARN,
      cal ? `top ${cal.topRef?.toFixed?.(0)}°` : 'none saved — the first run will capture it');

  // 11 · a demo path that does not need the camera
  add('fallback', 'Camera-free fallback', ctx.hasTraces ? Status.PASS : Status.WARN,
      ctx.hasTraces ? 'trace replay available' : 'record a trace so you can demo without detection');

  return out;
}

export function summariseChecks(results) {
  const c = { PASS: 0, WARN: 0, FAIL: 0, SKIP: 0 };
  for (const r of results) c[r.status] = (c[r.status] ?? 0) + 1;
  return {
    ...c,
    ok: c.FAIL === 0,
    verdict: c.FAIL ? 'NOT READY' : c.WARN ? 'READY, WITH WARNINGS' : 'READY',
  };
}

File: android/app/src/main/java/com/clashfit/engine/summary/Preflight.kt
package com.clashfit.engine.summary

/**
 * Pre-demo checklist: verify camera, inference, config, library, offline status, audio,
 * speech, haptics, storage, calibration, and camera-free fallback before a demo.
 */

enum class CheckStatus { PASS, WARN, FAIL, SKIP }

data class Check(
    val id: String,
    val label: String,
    val status: CheckStatus,
    val detail: String = "",
)

data class CheckSummary(
    val pass: Int,
    val warn: Int,
    val fail: Int,
    val skip: Int,
    val ok: Boolean,
    val verdict: String,
)

data class PreflightProbe(
    val running: Boolean = false,
    val fps: Float? = null,
    val inferMs: Float? = null,
    val poseConfig: PoseConfigProbe? = null,
    val exerciseCount: Int = 0,
    val isOnline: Boolean? = null,
    val audioEnabled: Boolean = false,
    val audioState: String? = null,
    val speechEnabled: Boolean = false,
    val speechVoice: String? = null,
    val hapticsAvailable: Boolean = false,
    val hapticsEnabled: Boolean = false,
    val sessionCount: Int = 0,
    val currentStreakDays: Int = 0,
    val calibrationExists: Boolean = false,
    val calibrationTopRef: Float? = null,
    val hasTraces: Boolean = false,
)

data class PoseConfigProbe(
    val version: Int = 1,
    val debugOverlay: Boolean = false,
)

/** Run the pre-demo checklist and return a list of checks. Never throws. */
fun preflight(probe: PreflightProbe): List<Check> {
    val out = mutableListOf<Check>()

    // 1. Camera running
    if (!probe.running) {
        out.add(Check("camera", "Camera running", CheckStatus.FAIL, "Start the camera before demoing."))
    } else if ((probe.fps ?: 0f) < 12) {
        out.add(Check("camera", "Camera running", CheckStatus.WARN, "only ${probe.fps?.toInt() ?: 0} fps"))
    } else {
        out.add(Check("camera", "Camera running", CheckStatus.PASS, "${probe.fps?.toInt() ?: 0} fps"))
    }

    // 2. Inference latency
    val infer = probe.inferMs
    when {
        infer == null -> out.add(Check("infer", "Inference latency", CheckStatus.SKIP, "no frames yet"))
        infer > 40 -> out.add(Check("infer", "Inference latency", CheckStatus.FAIL,
            "${infer.toInt()}ms — likely on CPU, not GPU"))
        infer > 22 -> out.add(Check("infer", "Inference latency", CheckStatus.WARN,
            "${infer.toInt()}ms, budget is 22"))
        else -> out.add(Check("infer", "Inference latency", CheckStatus.PASS, "${infer.toInt()}ms"))
    }

    // 3. Config loaded and debug off
    val poseConfig = probe.poseConfig
    when {
        poseConfig == null -> out.add(Check("config", "Config loaded", CheckStatus.FAIL, "no pose config"))
        poseConfig.debugOverlay -> out.add(Check("config", "Debug overlay OFF", CheckStatus.FAIL,
            "debugOverlay is true in pose.json — a jury will see it"))
        else -> out.add(Check("config", "Config loaded, debug off", CheckStatus.PASS, "v${poseConfig.version}"))
    }

    // 4. Exercise library
    out.add(Check("library", "Exercise library",
        if (probe.exerciseCount > 0) CheckStatus.PASS else CheckStatus.FAIL,
        "${probe.exerciseCount} exercises"))

    // 5. Offline
    when (probe.isOnline) {
        null -> out.add(Check("offline", "Works offline", CheckStatus.SKIP, ""))
        true -> out.add(Check("offline", "Works offline", CheckStatus.WARN,
            "currently online — put it in airplane mode and confirm the fight still runs"))
        false -> out.add(Check("offline", "Works offline", CheckStatus.PASS, "offline and running"))
    }

    // 6. Audio
    if (!probe.audioEnabled) {
        out.add(Check("audio", "Audio ready", CheckStatus.WARN,
            "not unlocked yet — click anything once"))
    } else {
        val state = probe.audioState ?: "unknown"
        val status = if (state == "running") CheckStatus.PASS else CheckStatus.WARN
        out.add(Check("audio", "Audio ready", status, state))
    }

    // 7. Speech
    val speechDetail = if (probe.speechEnabled)
        (probe.speechVoice ?: "default voice")
    else
        "no speech synthesis"
    out.add(Check("speech", "Speech available",
        if (probe.speechEnabled) CheckStatus.PASS else CheckStatus.WARN,
        speechDetail))

    // 8. Haptics
    out.add(Check("haptics", "Haptics",
        if (probe.hapticsAvailable) {
            if (probe.hapticsEnabled) CheckStatus.PASS else CheckStatus.PASS
        } else CheckStatus.SKIP,
        if (probe.hapticsAvailable) {
            if (probe.hapticsEnabled) "on" else "off"
        } else "no vibration motor"))

    // 9. Storage
    out.add(Check("storage", "Local storage", CheckStatus.PASS,
        "${probe.sessionCount} sessions, streak ${probe.currentStreakDays}"))

    // 10. Calibration
    val calDetail = if (probe.calibrationExists) {
        "top ${probe.calibrationTopRef?.toInt() ?: 0}°"
    } else {
        "none saved — the first run will capture it"
    }
    out.add(Check("calibration", "Calibration for this exercise",
        if (probe.calibrationExists) CheckStatus.PASS else CheckStatus.WARN,
        calDetail))

    // 11. Camera-free fallback
    out.add(Check("fallback", "Camera-free fallback",
        if (probe.hasTraces) CheckStatus.PASS else CheckStatus.WARN,
        if (probe.hasTraces) "trace replay available"
        else "record a trace so you can demo without detection"))

    return out
}

/** Summarise check results into verdict and counts. */
fun summariseChecks(results: List<Check>): CheckSummary {
    var pass = 0
    var warn = 0
    var fail = 0
    var skip = 0

    for (r in results) {
        when (r.status) {
            CheckStatus.PASS -> pass++
            CheckStatus.WARN -> warn++
            CheckStatus.FAIL -> fail++
            CheckStatus.SKIP -> skip++
        }
    }

    val ok = fail == 0
    val verdict = when {
        fail > 0 -> "NOT READY"
        warn > 0 -> "READY, WITH WARNINGS"
        else -> "READY"
    }

    return CheckSummary(pass, warn, fail, skip, ok, verdict)
}
